lib/unifex/code_generator.ex

```diff
@@ -2,55 +2,58 @@ defmodule Unifex.CodeGenerator do
   @moduledoc """
   Behaviour for code generation.
   """
+
   alias Unifex.Specs
 
   @type code_t :: String.t()
+  @type generated_code_t :: {header :: code_t, source :: code_t, generator :: module()}
 
   @callback identification_constant() :: String.t()
+  @callback subdirectory() :: String.t()
   @callback generate_header(specs :: Specs.t()) :: code_t
   @callback generate_source(specs :: Specs.t()) :: code_t
 
   @doc """
   Generates boilerplate code using generator implementation from `Unifex.CodeGenerators`.
   """
-  @spec generate_code(Specs.t()) :: [{header :: code_t, source :: code_t}]
+  @spec generate_code(Specs.t()) :: [generated_code_t()]
   def generate_code(specs) do
-    generators = get_generators(specs)
-
-    generators
-    |> Enum.map(fn {interface, generator} ->
+    for generator <- get_generators(specs) do
       header = generator.generate_header(specs)
       source = generator.generate_source(specs)
-      {header, source, interface}
-    end)
+      {header, source, generator}
+    end
   end
 
+  @spec get_generators(Specs.t()) :: [module()]
   defp get_generators(%Specs{name: name, interface: nil}) do
     {:ok, bundlex_project} = Bundlex.Project.get()
     config = bundlex_project.config
 
-    interfaces = [:natives, :libs] |> Enum.find_value(&get_in(config, [&1, name, :interface]))
+    generators =
+      [:natives, :libs]
+      |> Enum.find_value(&get_in(config, [&1, name, :interface]))
+      |> Enum.map(&bundlex_interface/1)
+      |> Enum.map(&interface_generator/1)
 
-    case interfaces do
+    case generators do
       [] -> raise "Interface for native #{name} is not specified.
         Please specify it in your *.spec.exs or bundlex.exs file."
-      _ -> interfaces |> Enum.map(&get_generator_module_name(&1))
+      _ -> generators
     end
   end
 
   defp get_generators(%Specs{interface: interfaces}) do
-    interfaces |> Bunch.listify() |> Enum.map(&get_generator_module_name(&1))
+    interfaces
+    |> Bunch.listify()
+    |> Enum.map(&interface_generator/1)
   end
 
-  defp get_generator_module_name(interface) do
-    module_name =
-      case interface do
-        :nif -> :NIF
-        :cnode -> :CNode
-        other -> other
-      end
+  @spec bundlex_interface(Bundlex.Native.interface_t()) :: Specs.interface_t()
+  defp bundlex_interface(:cnode), do: CNode
+  defp bundlex_interface(:nif), do: NIF
 
-    module = Module.concat(Unifex.CodeGenerators, module_name)
-    {interface, module}
-  end
+  @spec interface_generator(Specs.interface_t()) :: module()
+  def interface_generator(CNode), do: Unifex.CodeGenerators.CNode
+  def interface_generator(NIF), do: Unifex.CodeGenerators.NIF
```

Comment:
this way will be problematic in case of implementing more interfaces

Reply:
Agree, but my reasoning was that this solution would raise a `no function clause` in easily detectable place when we catch typo-style bug, instead of having us to debug effects of `Module.concat` enated atoms

Reply:
Than you can do `true = Code.ensure_loaded?(module)` on the concatenated module

Reply:
That sounds good, will add this check then